bottomless2/src/scheduler.rs

```diff
@@ -13,44 +30,342 @@ use crate::{
     /// namespace will be scheduled, until the `JobResult` is reported
+    #[tracing::instrument(skip_all)]
     pub fn schedule(&mut self) -> Option<Job<S>> {
-        todo!()
+        let (name, _) = self.queue.pop()?;
+        let requests = self
+            .requests
+            .get_mut(&name)
+            .expect("work scheduled but not requests?");
+        requests.in_flight = true;
+        let request = requests.requests.pop_front().unwrap();
+
+        tracing::debug!(job_id = request.id, "scheduled job");
+
+        let job = Job {
+            storage: self.storage.clone(),
+            request,
+        };
+
+        Some(job)
     }
 
     /// Report the job result to the scheduler. If the job result was a success, the request as
     /// removed from the queue, else, the job is rescheduled
-    pub fn report(&mut self, _job: JobResult<S>) {
+    #[tracing::instrument(skip_all, fields(req_id = result.job.request.id))]
+    pub async fn report(&mut self, result: JobResult<S>) {
         // re-schedule, or report new max durable frame_no for segment
-        todo!()
+        let name = result.job.request.request.namespace.clone();
+        let requests = self
+            .requests
+            .get_mut(&name)
+            .expect("request slot must exist");
+
+        requests.in_flight = false;
+
+        match result.result {
+            Ok(durable_index) => {
+                tracing::debug!("job success registered");
+                if self
+                    .durable_notifier
+                    .send((name.clone(), durable_index))
+                    .await
+                    .is_err()
+                {
+                    tracing::warn!("durability notifier was closed, proceeding anyway");
+                }
+            }
+            Err(e) => {
+                tracing::error!("error processing request, re-enqueuing");
+                // put it back at the front of the queue
+                requests.requests.push_front(result.job.request);
+            }
+        }
+
+        assert!(
+            self.queue.get_priority(&name).is_none(),
+            "there should be not enqueued jobs at this point"
+        );
+
+        if !requests.requests.is_empty() {
+            let first_id = requests.requests.front().unwrap().id;
+            self.queue.push(name, Reverse(first_id));
+        } else {
+            self.requests.remove(&name);
+        }
     }
 
     /// Returns true if the scheduler is empty, that is, there are no scheduler requests, and
     /// no not-scheduled request: iow, it's empty.
     pub fn is_empty(&self) -> bool {
-        todo!()
+        self.requests.is_empty()
     }
 
     /// Scheduler has work to do. Calling `schedule` after this method must always return some job
     pub fn has_work(&self) -> bool {
-        todo!()
+        !self.queue.is_empty()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::path::PathBuf;
+
```

Comment:
```suggestion
```

Reply:
this is on purpose, I separate std imports, from extern, for intern. It's a common style convention (so much so that RA, and rustfmt does it automatically: https://github.com/rust-lang/rustfmt/issues/2979)